Allow skipping cache saving in decode phase

yoo-kumaneko opened the pull request:
Credit to @TeenSpirit1107

maobaolong commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -241,6 +244,11 @@ def update(
             new_block_ids = new_block_ids[0]
         self.allocated_block_ids.extend(new_block_ids)
 
+        # When a request is scheduled again, and the number of new tokens is 1 (exluding chunked prefill)
```
You can mark this as TODO or FIXME comment

yoo-kumaneko replied:
done

maobaolong reviewed (commented):
Left a minor comment @yoo-kumaneko  otherwise, it looks good. @YaoJiayi @ApostaC Would you like to take another look?

yoo-kumaneko commented:
This is a cleaner implementation. Needs reviewing.

ApostaC commented:
Will take a look today

ApostaC commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -284,12 +295,19 @@ def from_request_tracker(
         # 2. number of unsaved tokens is not reached the chunk boundary
+        # 3. if save_decode_cache is False in decode phase
+
+        # Skip save if we're in decode phase and save_decode_cache is False
+        save_decode_cache = lmcache_connector.config.save_decode_cache
```
Since we are introducing `save_decode_cache` back, one quick check: is the default config for save_decode_cache True or False?

My proposal is make it `True` by default and users can turn it off explicitly. What do you guys think? @yoo-kumaneko @maobaolong @YaoJiayi

maobaolong replied:
@ApostaC I'm not sure whether `True` or `False` is the better default value, but there are some consideration items:
- The original behavior of v0 is `False`, it means we do not support save decoded code by default before. 
- There are more cost if we save decoded kvcaches, and maybe the decoded kvcaches can not be reused for some scenarios.

yoo-kumaneko replied:
@ApostaC  If the current default behavior is True, let's keep it this way. We don't want to surprise users with inconsistent behavior.

ApostaC replied:
Yeah, I think the current behavior is `True`, and that's the reason of my proposal.

yoo-kumaneko replied:
@ApostaC Got it done. The default value is now True in `lmcache/v1/config.py`, while the default value in `lmcache/config.py` stays unchanged (which is False). Do you think we should modify `lmcache/config.py`?

ApostaC replied:
We don't have to modify `lmcache/config.py`. That code is deprecated and we should remove it soon.

yoo-kumaneko replied:
@ApostaC Could you give me some hints on adding a test to the CI? It seems like the current CI test is only a basic functional test. Should I add the save_decode_cache argument to the config and check that it doesn’t crash, or is there something else I should focus on?

ApostaC replied:
I think #1000 introduced end-to-end CI test with vLLM. 
@Shaoting-Feng Can you give some instructions about how to add a new vLLM integration test? Thanks!

Shaoting-Feng replied:
You can add the vLLM and LMCache configurations to `scripts/vllm-integration-tests.sh` and start a new test.

yoo-kumaneko replied:
I've added a new test to the `vllm-integration-tests.sh` file. Could you please review it and let me know if it looks good?

ApostaC commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -257,6 +266,7 @@ class ReqMeta:
 
     @staticmethod
     def from_request_tracker(
+        lmcache_connector: "LMCacheConnectorV1Impl",
```
I think it's a bit heavy to pass the whole `lmcache_connector` in. And the code assumes the existence of `lmcache_connector.config.save_decode_cache`. 

Would be better if we only pass in a flag of `save_decode_cache`.

yoo-kumaneko replied:
Makes sense. Will do.

ApostaC reviewed (commented):
LGTM in general. Just some minor comments.

Also, would be great to share some experiment results to make sure it works. We have introduced the vllm integration test into our CI (#1000), so maybe it's a good place to add some basic/quick test there.

maobaolong commented:
@ApostaC As discussed offline, the existing save_decode_cache maybe cannot be reused, besides, the `save_decode_cache` is not the expected behavior of `v1`, do you think it would be a little better if we make it default to False?

ApostaC commented:
> @ApostaC As discussed offline, the existing save_decode_cache maybe cannot be reused, besides, the `save_decode_cache` is not the expected behavior of `v1`, do you think it would be a little better if we make it default to False?

Sounds good to me!

Also, is this PR ready for review again?

yoo-kumaneko commented:
> > @ApostaC As discussed offline, the existing save_decode_cache maybe cannot be reused, besides, the `save_decode_cache` is not the expected behavior of `v1`, do you think it would be a little better if we make it default to False?
> 
> Sounds good to me!
> 
> Also, is this PR ready for review again?

Yes. I've changed the default option for `save_decode_cache` to False, and added a CI test for it. It's now ready for review again.

Shaoting-Feng commented on .buildkite/scripts/vllm-integration-tests.sh:
```diff
@@ -29,6 +29,15 @@ SERVER_WAIT_TIMEOUT=180
 # UTILITIES #
 #############
 
+cleanup_container() {
```
Is it possible to utilize `.buildkite/scripts/docker-cleanup.sh`?

Shaoting-Feng commented on .buildkite/scripts/vllm-integration-tests.sh:
```diff
@@ -113,6 +122,72 @@ run_lmcache_vllmopenai_container() {
+            --gpu-memory-utilization '0.5' \
+            --enforce-eager)
+    fi
+    buildkite-agent meta-data set "docker-CID" "$CID"
```
This may overwrite the `docker-CID` meta-data.

Shaoting-Feng commented on .buildkite/scripts/vllm-integration-tests.sh:
```diff
@@ -150,6 +225,32 @@ test_vllmopenai_server_with_lmcache_integrated() {
+        cleanup 1
+        exit 1
+    else
+         echo "Model prompt request from OpenAI API server with save decode cache succeeded"
```
This still doesn’t verify whether saving the decode cache was executed. But I do not have any good idea in mind.

yoo-kumaneko replied:
Me neither. It seems that the main challenge here is the lack of test infrastructure. The current tests we have are limited to launch tests, which don’t cover function-level verification. Would it be possible to raise a separate issue specifically for adding tests, so we can keep this PR focused and move forward? I’ve tested this locally, and it works as expected.

Shaoting-Feng replied:
> Would it be possible to raise a separate issue specifically for adding tests

That makes sense. Thank you for your PR!

Commit:
Add _save_decode_cache in adapter

Signed-off-by: rigginschen <rigginschen@tencent.com>

maobaolong reviewed (approved):
@yoo-kumaneko Sorry for ignoring this PR the past weeks, it looks good to me, and I have used this for my single node test already.

yoo-kumaneko committed:
Merge branch 'dev' into skip-saving-decode-cache

maobaolong committed:
Merge branch 'dev' into skip-saving-decode-cache

maobaolong commented:
@TeenSpirit1107 Thank you very much for your great effort on https://github.com/LMCache/LMCache/pull/916 , and thanks @yoo-kumaneko for the follow-up work by this PR.

Merged it.
